fetch.py

```diff
@@ -7,21 +7,154 @@
 
 
 """
+from datetime import datetime
 import sys
 import os
 import json
-import logging
-import logging.handlers
 import hashlib
 import requests
 import openstack
 
-logger = logging.getLogger(__name__)
+
+
+class ImgBuildLogger:
+    """
+    Set value for exitting the program and handle logging
+    0 = OK
+    1 = Warning
+    2 = Error
+    """
+
+    # Rotates logs with this
+    MAX_LOG_SIZE = 5000
+
+    def __init__(self) -> None:
+        """
+        Init to ok
+        """
+        self._code = 0
+        self._log = {
+            "start_timestamp": str(datetime.now()),
+            "PID": os.getpid(),
+            "cloud": None,
+            "current": {},
+            "deprecated": {}
+        }
+        self._cur_write = self._log["current"]
+        self._cur_img = None
+
+
+    def start_deprecated(self) -> None:
+        """
+        Starts to fill log.deprecated instead of log.current
+        """
+        self._cur_write = self._log["deprecated"]
+
+
+    def set_image(self, image: str) -> None:
+        """
+        Sets the current image that's being written
+        """
+        self._cur_img = image
+        self._cur_write[self._cur_img] = self._cur_write.get(self._cur_img, {})
+        print(f"=== {image} ===")
+
+    def set_cloud(self, cloud: str) -> None:
+        """
+        Sets the cloud variable for the log
+        """
+        self._log["cloud"] = cloud
+
+
+    def _write_log(self, msg: str, level: str, important: bool = False) -> None:
+        """
+        Helper function that writes the messages into the log
+        """
+        message = {
+            "level" : level,
+            "content" : msg
+        }
+
+        if important:
+            message["important"] = True
+
+        self._cur_write[self._cur_img][str(datetime.now())] = message
+        print(f"{level}: {msg}")
+
+
+    def warning(self, msg: str) -> None:
+        """
+        Set code to warning and log the message
+        """
+        self._code = max(self._code, 1)
+        self._write_log(msg, "WARNING")
+
+    def error(self, msg: str) -> None:
+        """
+        Set code to error and log the message
+        """
+        self._code = 2
+        self._write_log(msg, "ERROR")
+
+
+    def info(self, msg: str, important: bool = False) -> None:
+        """
+        Same as logger.info
+        """
+        self._write_log(msg, "INFO", important)
+
+
+    def write_log(self, log_file: str) -> None:
+        """
+        Writes the current log buffer into a file and rotates logs
+        """
+        try:
+            with open(log_file, "r", encoding="utf-8") as f:
+                current_log_data= json.load(f)
+
+        except FileNotFoundError:
+            current_log_data = []
+
+        except json.decoder.JSONDecodeError:
+            current_log_data = []
+
+
+        current_log_data.append(self._log)
+
+        current_log_data = current_log_data[-self.MAX_LOG_SIZE:]
```

Comment:
it might be easier to forget about the log rotation in fetch.py (considering that it is not provided ready-for-use anymore by logging module) and offload this task to the other repo of yours (ansible + configs) configuring logrotate to take care of it

Reply:
here some examples:
https://github.com/CSCfi/ansible-role-rally/blob/master/tasks/configure_rally.yml#L80
https://github.com/CSCfi/ansible-role-rally/blob/master/files/rallyverifiers-logrotate